core/templates/dev/head/pages/exploration-editor-page/improvements-tab/services/improvement-modal.service.ts

```diff
@@ -48,6 +48,177 @@ angular.module('oppia').factory('ImprovementModalService', [
       ThreadStatusDisplayService, UrlInterpolationService,
       UserExplorationPermissionsService, UserService) {
     return {
+      /**
+       * Opens the modal for displaying playthrough actions.
+       * @param {Playthrough} playthrough.
+       * @param {int} playthroughIndex.
+       */
+      openPlaythroughModal: function(playthrough, playthroughIndex) {
+        $uibModal.open({
+          templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+            '/pages/exploration-editor-page/statistics-tab/templates/' +
+            'playthrough-modal.template.html'),
+          backdrop: true,
+          controller: [
+            '$scope', '$uibModalInstance', 'AlertsService',
+            'LearnerActionRenderService',
+            'ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS',
+            'MAX_UNRELATED_ACTIONS_PER_BLOCK',
+            function(
+                $scope, $uibModalInstance, AlertsService,
+                LearnerActionRenderService,
+                ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS,
+                MAX_UNRELATED_ACTIONS_PER_BLOCK) {
+              $scope.playthroughIndex = playthroughIndex;
+
+              // This is the index of the first action in the range of actions
+              // displayed. When this index reaches 0, we've displayed all the
+              // learner actions.
+              let indexOfFirstDisplayedAction = playthrough.actions.length - 1;
+
+              /**
+               * Identifies whether all learner actions have been displayed and
+               * no further expansion is possible.
+               * @returns {boolean}
+               */
+              $scope.canExpandActions = function() {
+                return indexOfFirstDisplayedAction !== 0;
+              };
+
+              /**
+               * Iterates through the learner actions and expands the set of
+               * actions to be displayed. If all learner actions have been
+               * handled, no more expansion is possible.
+               */
+              let expandActions = function() {
+                let i;
+                let previousStateName = null;
+                if (indexOfFirstDisplayedAction ===
```

Comment:
Given that the index defined above starts at N - 1 and goes backwards, surely this will never be the case?

Reply:
Right. My bad. Actually this case is not required. The first time we enter this function, its okay for the previousStateName to be null. So we only need the else condition.